# <font color=red> Solutions for</font> Exercises for Lab 02: <font color=blue>Version 0.9 Beta</font>
Welcome to the Exercises for Lab 2! Completion of this will net you TWO extra credit points! There might be more around here, though.

## An intro to Importing Libraries

In [1]:
import numpy as np
import matplotlib
import math as m
from datascience import *

The above 'importing' of these libraries, as they are called, is akin to the unlocking of a whole new set of functions, methods, and other workings to be used in your code. If you do not include and run these imports, you won't have access to these tools.

These additional tools' appearances depend on how you import them. Specifically, their functions typically appear with the name of their library before it. For example, since we imported numpy as np, its functions have 'np' before it. See below for examples.

In [5]:
np.arange(5)

array([0, 1, 2, 3, 4])

In [6]:
arange(5)

NameError: name 'arange' is not defined

In [7]:
import numpy
numpy.arange(5)

array([0, 1, 2, 3, 4])

See how, since we imported numpy again but with no clarification on its name, we can thus use its full name 'numpy' before the function name. (Since we also imported it as 'np' earlier, the 'np.arange' format still works.)

This means you can technically use anything under the sun as the name of a library.

In [8]:
import numpy as tomato
tomato.arange(5)

array([0, 1, 2, 3, 4])

If you really REALLY want to import the library without its name, it requires a bit of formatting and the use of an asterisk. This is super risky, however, and should be used only if there are very few functions in the library (like in the Datascience module.)

In [9]:
from numpy import *
arange(5)

array([0, 1, 2, 3, 4])

This 'from (method) import (function)' can be used to render specific functions without their module name. If I run the cell below this without having done what I did in the previous cells, only arange would be able to be used without its module name, and all other functions would NOT be imported.

In [10]:
from numpy import arange
arange(5)

array([0, 1, 2, 3, 4])

You can find additional info on any library by typing in any of its defined names (in this case np, numpy, and tomato are all available since we assigned them all to this numpy library earlier) with a question mark.

In [11]:
tomato?

Type:        module
String form: <module 'numpy' from '/opt/conda/lib/python3.11/site-packages/numpy/__init__.py'>
File:        /opt/conda/lib/python3.11/site-packages/numpy/__init__.py
Docstring:  
NumPy
=====

Provides
  1. An array object of arbitrary homogeneous items
  2. Fast mathematical operations over arrays
  3. Linear Algebra, Fourier Transforms, Random Number Generation

How to use the documentation
----------------------------
Documentation is available in two forms: docstrings provided
with the code, and a loose standing reference guide, available from
`the NumPy homepage <https://numpy.org>`_.

We recommend exploring the docstrings using
`IPython <https://ipython.org>`_, an advanced Python shell with
TAB-completion and introspection capabilities.  See below for further
instructions.

The docstring examples assume that `numpy` has been imported as ``np``::

  >>> import numpy as np

Code snippets are indicated by three greater-than signs::

  >>> x = 42
  >>> x = x + 1

U

You can also do help(library name), but I do caution you against it because of just how comprehensive it is. Instead, you can see that they provide external resources and other assistance with the library in the cell output we just saw.

## Strategy of Coding and the Learning Process
At its core, how you approach learning in any subject will affect your proficiency down the line. I personally believe that there is little to no difference between studying a language (e.g. English, French, Russian, etc.) and studying a STEM subject (e.g. Chemistry, Math, Physics), and given that coding (at least in this course) uses aspects of both ends (syntax/grammar vs numbers/data), it seems more than natural that one should adopt a process of learning that is similar to them.

Now what do I mean when I talk about the 'process of learning' of studying languages or STEM subjects? Most prominently, I mean that one should think critically above all else. So many people (including past me) believe that studying is all about brute force memorization, but that just strains your mind and doesn't amount to much in the long term. It’s much better to think of all aspects of code as the toolbox that you use for some ulterior purpose. In this sense, the main goal of critical thinking is to chiefly know what each function or method does or can do, and thus what they cannot do (the extent of their ability).

As a basic example, you know that a hammer works best for when you want to put a nail in something, and you know that a screwdriver works best for when you want to put a screw in something. Being a beginner at coding is equivalent to not being knowledgeable as to what tools to use for what purpose. (Do I use the hammer for screws? Do I use the screw for nails?) However, as you work with it more and more, you will come to make those connections. Soon, you will be able to tackle more challenging problems, for example: How do you hang a picture on a wall? Well, you need to mark out the position that you want to hang it, you need to drive a nail into that wall (using a hammer), you need to put the picture in a frame, and you need to hang that frame on the wall. This may seem obvious to the point of redundancy, but that’s on purpose; the functions and methods of code are not nearly as obvious in their purposes as real-life examples like hammers and screwdrivers are, but they do function in the same sort of manner and scope.

(Apologies for the giant wall of text; if you are reading this, it probably means that you were curious enough (or desperate enough) to read through all the stuff I had to say here. As thanks, I give you another extra extra credit credit opportunity. Please create another Markdown cell below this, and in that cell please give feedback on how helpful this above advice is. Although I believe fervently in it, I have yet to put it use in a Data Science class. Doing this will net you 1 more extra extra credit credit point.)

However, you need to assemble your toolbox first. This is best done through pages in Jupyter Notebook that you can easily access (see the tables just below for examples), but it can be done in Word if you so desire. Whenever you encounter a new function/method/whatever-have-you, no matter what lab or other source, list the following:
* The function or method in the format that you would input, including the arguments, both necessary and optional.
* What the function or method does, and what it outputs.
* The library that it is a part of (numpy, datascience, matplotlib, etc.).
* Any additional things to remember
* Perhaps an example of you using it (not required if you can understand the function without it).

Even the act of seeing in-lab examples may not paint the full picture. Additionally, a big part of coding is to be able to find new things that can speed your life up and make things easier. To see all aspects of a function/method, do any of the following.
* If it’s part of a library, you can get tons of information by searching for in the library's API reference. You can very quickly get overwhelmed (since ALL functions/methods of the library are there) but it does give you the fullest picture of what that function/method does. Here are some links:
    * [Numpy API reference](https://numpy.org/doc/2.3/reference/index.html#reference)
    * [Datascience API reference](https://www.data8.org/datascience/reference-nb/datascience-reference.html)
    * [Matplotlib.Pyplot API reference](https://matplotlib.org/stable/api/pyplot_summary.html)
    * All of these have some sort of Search button/bar that allows you to very easily search for the function/method you seek. (don't wade through the forest of functions for hours on end!)
* If you want to know more about the entire library within the Jupyter Notebook, do `help(Library)`. (this is pretty dense)
* You can also do `FunctionName?` or, if in a library, `Library.FunctionName?`
* Methods are in the form `DataType.MethodName?`, but some do work without the datatype part. (You may need to experiment a bit.) See below for examples.


In [12]:
list.remove?
# this remove method operates on a list, hence why list is at the beginning.

Signature: list.remove(self, value, /)
Docstring:
Remove first occurrence of value.

Raises ValueError if the value is not present.
Type:      method_descriptor

In [13]:
split?
# in code, this would usually be in the form list.split(...) but apparently it does work fine without it.

Signature: split(ary, indices_or_sections, axis=0)
Docstring:
Split an array into multiple sub-arrays as views into `ary`.

Parameters
----------
ary : ndarray
    Array to be divided into sub-arrays.
indices_or_sections : int or 1-D array
    If `indices_or_sections` is an integer, N, the array will be divided
    into N equal arrays along `axis`.  If such a split is not possible,
    an error is raised.

    If `indices_or_sections` is a 1-D array of sorted integers, the entries
    indicate where along `axis` the array is split.  For example,
    ``[2, 3]`` would, for ``axis=0``, result in

      - ary[:2]
      - ary[2:3]
      - ary[3:]

    If an index exceeds the dimension of the array along `axis`,
    an empty sub-array is returned correspondingly.
axis : int, optional
    The axis along which to split, default is 0.

Returns
-------
sub-arrays : list of ndarrays
    A list of sub-arrays as views into `ary`.

Raises
------
ValueError
    If `indices_or_sections` is given as an

## Lists and Arrays

Lists and Arrays are extremely similar in how they work. Being two of the most-used data types in this program, it is imperative that you get a hold on what is a list vs what is an array, and how they differ in appearance, use, and other aspects.

Note: a component of a list or array is called an element. The placement of an element in a list or array, with the first being designated as 0, the second as 1, and so on, is the Index of the element.

Any and all parts of the code intended to be customizable as befit the situation (e.g. what list are you trying to append stuff to? what variable is this stuff defined as?) has a capital letter (Functions/Methods like these do not use capital letters).

|Descriptor|List|Array|
|--|--|--|
|How to make it:|simply type in the list elements as following:`ListName = [Element1, Element2, Element3, ...]`|use the datascience module command `ArrayName = datascience.make_array(Element1, Element2, Element3, ...)` or, if it is a sequence of numbers, `ArrayName = np.arange(Start, End, Step)` (Start and end are optional; omit them if you are going from 0 and up by 1)|
|Composition:|accepts most if not all data types wherever|accepts most of not all data types, but a single array must have only one data type|
|Location in a Table (spoiler for Lab 3!):|a single row of a table|a single column of a table, minus the header row|


List of commands:

|Effect|List|Array|
|--|--|--|
|Getting the Index of an Element|`ExistingList.index(Element)` (this only gives you the first occurrence as an integer datatype)|`np.where(ExistingArray == Element)` (gives you an array of all the indexes the target Element is found)|
|Getting an Element from its Index|`ExistingList[IndexOfElement]`|`ExistingArray[IndexOfElement]`|
|Adding an element to it|`ExistingList = ExistingList + [NewElement]` or `ExistingList.append(NewElement)`|`ExistingArray = np.append(ExistingArray, NewElement)`|
|Removing an element|`ExistingList.remove(Element)` or `del ExistingList[IndexOfElement]` or `ExistingList.pop(IndexOfElement)`|`ExistingArray = ExistingArray.delete(IndexOfElement)`|
|Changing a specific element|`ExistingList[IndexOfTargetElement] = NewElement`|`ExistingArray[IndexOfTargetElement] = NewElement`|
|Modifying all elements by the same operation/function/method (this works for all of these three; I'll use the abbreviation FOM for where you should put this) (There are a few examples below)|`[FOM(x) for x in List]`|`[FOM(x) for x in Array]` (output is a list) OR `FOM(Array)` (output is an array)|
|Exclusive for Lists: Modifying/Filtering with a Conditional|`[FOM(x) for x in List if ...]` (`...` is where you put the conditional statement, like `x == 5` or `x <= 3`) OR `[x for x in List if ...]` to get only the elements of the list that match that conditional (filtering)|Replace the list with an array to filter it. However, the output will be an list.|

|Effect|Function|
|--|--|
|average|`np.average(listOrArray)`|
|maximum|`max(listOrArray)`|
|minimum|`min(listOrArray)`|
|sum|`sum(listOrArray)`|
|length (number of elements)|`len(listOrArray)`|

|Conditional|Meaning|
|--|--|
|`x == a`|Is x equal to a?|
|`x != a`|Is x not equal to a?|
|`x > a`|Is x greater than a?|
|`x < a`|Is x less than a?|
|`x >= a`|Is x greater than or equal to a?|
|`x <= a`|Is x less than or equal to a?|
|`x in a`|Is x contained within a (in the situation where a is a list/array/string)?|

### Lists: Element Manipulation
Lists are more data oriented. They typically treat their elements the same regardless of data type. Multiplying by two doesn't double the values of any floats or integers, but instead repeats the composition of the list again.

In [14]:
list1 = [52, 23, 66]
list1*2

[52, 23, 66, 52, 23, 66]

Concatenation and Mathematical operations: You must specify that you are doing this operation to each element in the list.

In [15]:
list2 = ['hope', 'delight', 'doubt']
[i+'ful' for i in list2]

['hopeful', 'delightful', 'doubtful']

In [16]:
[i*2 for i in list1]
# example of a math operation being done on all elements

[104, 46, 132]

In [17]:
[len(x) for x in list2]
# example of a function being done on all elements

[4, 7, 5]

In [18]:
list3 = ['Panama Canal', 'Banana', 'Appalachians']
[x.split('a') for x in list3]
# example of a method being done on all elements

[['P', 'n', 'm', ' C', 'n', 'l'],
 ['B', 'n', 'n', ''],
 ['App', 'l', 'chi', 'ns']]

### Arrays: Element Manipulation
It might help to think of arrays as being more numerically oriented. Any mathematical operations on arrays made of floats or integers behave as though the operation is being done on each element individually.

Concatenation and Mathematical Operations: Since Arrays treat numbers like math, you can do any mathematical operations to the array itself, and it will apply to all terms inside.

In [19]:
array1 = make_array(52, 23, 66)
array1*2

array([104,  46, 132])

You can still use the same thing for lists as for arrays, but the output here is a list.

In [20]:
[x**2 for x in array1]

[2704, 529, 4356]

Concatenation works slightly differently, but it is still noticeably easier on arrays than on lists.

In [21]:
array2 = make_array('animate', 'corroborate', 'disguise', 'orientate')
array2 = np.char.add(array2,'d')
array2

array(['animated', 'corroborated', 'disguised', 'orientated'],
      dtype='<U12')

### Arrays as containing only One Datatype
The Numpy Array's main difference from lists is that it limits itself to only one datatype, and any new datatype added to it will cause the array's elements to modify themselves to match a common datatype.

Example: An array made of arrays.

In [22]:
from datascience import *
test1 = make_array('blah', 'blah', 'blah')
test2 = make_array('stuff', 'stuff', 'stuff')
arr = make_array(test1, test2)
print('The First Element here is a', str(type(arr[1])))
print('The Length of this array is', len(arr))
arr

The First Element here is a <class 'numpy.ndarray'>
The Length of this array is 2


array([['blah', 'blah', 'blah'],
       ['stuff', 'stuff', 'stuff']], dtype=object)

If you add another element of a different datatype, the array will accept it, but it will modify its elements such that all of them are the same datatype.

Exhibit 1: arr had arrays previously. If I add a string, the arrays inside are broken apart into their separate strings.

In [23]:
arr = np.append(arr, 'this is a string, not an array.')
print('The First Element here is a', str(type(arr[1])))
print('The Length of this array is', len(arr))
arr

The First Element here is a <class 'str'>
The Length of this array is 7


array(['blah', 'blah', 'blah', 'stuff', 'stuff', 'stuff',
       'this is a string, not an array.'], dtype=object)

Exhibit 2: making an array consisting of a string, an integer, and a float turns all elements into strings. Note that it does this because a string can't reliably be converted into floats or integers the same way floats and integers can be converted into strings.

In [24]:
test3 = make_array('this is a string.', 3, 5.0)
print('The Integer 3 became a', type(test3[1]))
test3

The Integer 3 became a <class 'numpy.str_'>


array(['this is a string.', '3', '5.0'],
      dtype='<U32')

## Strings
Strings are one of the three main datatypes alongside floats and integers. While those two are fairly straightforward, the string is more complex because it behaves not like a single character or an inseparable cluster thereof, but rather a sequence in the same way that lists and arrays are sequences of characters.

See the table below for a list of functions.

|Function|String|
|--|--|
|How to get the Element of a certain Index|`existingString[indexOfElement]`|
|How to add Strings together (concatenation)|`newString = existingString+existingString`|
|How to replace elements of a string with another|`existingString = existingString.replace(oldElement, newElement, indexOfoldElement(optional))` (if the index is not given, all instances of the oldElement will be replaced.)|
|How to delete elements from a string|use the above .replace function, but with the newElement being simply an empty quotation|
|How to take a certain region of the string|`newString = existingString.split(boundaryElement)[indexOfTargetSection]` or `newString = existingString[startingIndex:finishingIndex]` (leave the starting or finishing index blank if going from the beginning or going to the end)|
|How to calculate a mathematical expression in String format|`eval(mathExpressionAsString)`|
|How to split up a string|`existingString.split(splitBoundaryAsString)` (this consumes the split boundary and returns a list of the resultant pieces)|

In [24]:
apple = 'apple'
apple = apple.replace('p', 'g')
apple

'aggle'

In [25]:
One_and_a_half = '2*3/4'
eval(One_and_a_half)

1.5

## The Actual Exercises
Go through these tasks with the intention of doing things step by step, and refer back to the toolbox up above. 

### Exercise 1: Odd Numbers

1.1: You can express any odd number as the difference of the two squares of consecutive integers. For example, the difference between 4<sup>2</sup> and 5<sup>2</sup> is 9. Use this fact to create a list of the first 100 positive odd numbers; start from an array of the first 100 numbers (both odd and even). Store your result under the variable name `e1_1`.

Hints:
* What's a super concise way of generating a hundred numbers?
* If any number in the array is `x`, how do you represent the number that is one above it?

In [2]:
ints = np.arange(1,101)
e1_1 = [x**2-(x-1)**2 for x in ints]
print(e1_1)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 193, 195, 197, 199]


1.2: You can also express any odd number as the formula `2n-1`, where n is a whole number integer. Use this to generate a list of the first 100 positive odd numbers, and store that under `e1_2`.

In [3]:
e1_2 = [2*x-1 for x in ints]
print(e1_2)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 193, 195, 197, 199]


1.3: Of course, you can always take the simple route and filter a list of numbers based on oddness/evenness. Use a list of the first 200 integers to get a list of the first 100 positive odd numbers, and store that under `e1_3`. Hint: when using the `if` filtering, the double equals signs `==` means that the two sides are alike (in contrast to a single equals sign, which is for assigning values to variables).

In [4]:
ints200 = np.arange(1,201)
e1_3 = [x for x in ints200 if x%2 == 1]
print(e1_3)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 101, 103, 105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129, 131, 133, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 193, 195, 197, 199]


1.4: And perhaps the easiest way, you can just use `np.arange` and a step size of two to generate the odd numbers.

In [5]:
e1_4 = np.arange(1,200,2)
print(e1_4)

[  1   3   5   7   9  11  13  15  17  19  21  23  25  27  29  31  33  35
  37  39  41  43  45  47  49  51  53  55  57  59  61  63  65  67  69  71
  73  75  77  79  81  83  85  87  89  91  93  95  97  99 101 103 105 107
 109 111 113 115 117 119 121 123 125 127 129 131 133 135 137 139 141 143
 145 147 149 151 153 155 157 159 161 163 165 167 169 171 173 175 177 179
 181 183 185 187 189 191 193 195 197 199]


In [6]:
type(e1_4)

numpy.ndarray

In [7]:
# Use this cell to check whether the three lists are the same
if type(e1_1) != list and type(e1_2) != list and type(e1_3) != list and type(e1_4) != numpy.ndarray:
    print('Did you do all of the three exercises above?')
elif e1_1 == e1_2 and e1_2 == e1_3 and e1_3 == list(e1_4):
    print('All four lists are the same! Good Job!')
else:
    print('oops! there seems to be some incongruity...')

All four lists are the same! Good Job!


### Exercise 2: Prime Numbers
2.1: The square of any Prime Number is one more than a multiple of 24. For example, 5<sup>2</sup> is 25, which is `24*1+1`, and 7<sup>2</sup> is 49, which is `24*2+1`. In essence, p<sup>2</sup> = 24n+1, where n is any integer. Solve this for p, and then generate the first 19 primes in this formula by starting with an array of 200 integers (you'll see why we need so many!) from 1 to 200. Make sure to avoid 0 in that initial array.

In [8]:
# I use the ints200 variable as defined above
p24 = [(24*x+1)**.5 for x in ints200]
print(p24)

[5.0, 7.0, 8.5440037453175304, 9.8488578017961039, 11.0, 12.041594578792296, 13.0, 13.892443989449804, 14.730919862656235, 15.524174696260024, 16.278820596099706, 17.0, 17.691806012954132, 18.357559750685819, 19.0, 19.621416870348583, 20.223748416156685, 20.808652046684809, 21.377558326431949, 21.931712199461309, 22.472205054244231, 23.0, 23.515952032609697, 24.020824298928627, 24.515301344262525, 25.0, 25.475478405713993, 25.942243542145693, 26.40075756488817, 26.851443164195103, 27.294688127912359, 27.730849247724095, 28.160255680657446, 28.583211855912904, 29.0, 29.410882339705484, 29.816103031751148, 30.215889859476256, 30.610455730027933, 31.0, 31.384709652950434, 31.764760348537184, 32.140317359976393, 32.511536414017719, 32.878564445547191, 33.241540277189323, 33.600595232822883, 33.955853692699286, 34.307433596816885, 34.655446902326908, 35.0, 35.341194094144583, 35.679125549822544, 36.013886210738214, 36.345563690772494, 36.674241641784491, 37.0, 37.322915213043039, 37.6430604

You may realize that not all of the terms in there are whole numbers. Try and filter them such that only the whole numbers remain. Name the filtered list `e2_1`.

In [9]:
e2_1 = [x for x in p24 if x % 1 == 0]
print(e2_1)

[5.0, 7.0, 11.0, 13.0, 17.0, 19.0, 23.0, 25.0, 29.0, 31.0, 35.0, 37.0, 41.0, 43.0, 47.0, 49.0, 53.0, 55.0, 59.0, 61.0, 65.0, 67.0]


Make sure to store your list as `e2_1`, then use the following cell to check against a preset list of primes. Are there any imposters? How can you get rid of them?

In [10]:
# just run this
preset_primes = [5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67]
trouble_spots = []
for i in e2_1:
    if np.count_nonzero([n == i for n in preset_primes]) == 1:
        pass
    else:
        trouble_spots.append(i)
trouble_spots

[25.0, 35.0, 49.0, 55.0, 65.0]

2.2: In Prime Numbers, a classic is what is known as Euler's Lucky Numbers (or Caboose Numbers), which are integers p such that $x^2-x+p$ gives prime numbers when x is any integer from 0 to p-1. The most famous Euler Lucky Number is 41; that is, for all integers from 0 to 40, this expression $x^2-x+41$ gives primes.

Since this is so similar to problems like Exercise 1, I challenge you to get the 41 primes that can come from the expression $x^2-x+41$ in just ONE line of code. Good luck!

In [11]:
e2_2 = [x**2-x+41 for x in np.arange(1,41)]
print(e2_2)

[41, 43, 47, 53, 61, 71, 83, 97, 113, 131, 151, 173, 197, 223, 251, 281, 313, 347, 383, 421, 461, 503, 547, 593, 641, 691, 743, 797, 853, 911, 971, 1033, 1097, 1163, 1231, 1301, 1373, 1447, 1523, 1601]


2.3: There are three other caboose numbers smaller than 41 and bigger than 3 that also have this property. Build code to search for them. I will provide a helpful prime reference list to check the generated primes against. Since the lower numbers are a bit floppy, start searching from 4 onwards.

Here are some hints:
* Since this involves lists within lists, you'll need to do plenty of list operations inside list operations. Keep track of what each list contains!
* You're going to be comparing the contents of each sub-list against the reference list. Notably, you'll need to find some way to filter just those that are inside the reference list from those that are NOT.
* You're also going to be comparing the filtered list (that with just the primes contained in the reference list, and without those that are non-primes) with your original non-filtered list. You'll need to be very creative here; how can you compare each term of the two lists such that only those that have not at all changed are left?
* Indices provide a very helpful way for you to get the position of any component in a list. Before you go off taking away terms in a list, note that the removal of any term in a list shifts the position of all terms after it. Perhaps this might not be to your benefit.
* Since you're starting from 4, the indices will not accurately match the number you're analyzing. How can you re-shift the indices to restore that match? (For example, the terms of x^2-x+4 are in a sublist that is the very first term of the larger list, and thus have index 0, and NOT index 4.)

In [12]:
testing = np.arange(4, 41)
potentials = [[x**2-x+y for x in np.arange(1,y)] for y in testing]
print(potentials)

[[4, 6, 10], [5, 7, 11, 17], [6, 8, 12, 18, 26], [7, 9, 13, 19, 27, 37], [8, 10, 14, 20, 28, 38, 50], [9, 11, 15, 21, 29, 39, 51, 65], [10, 12, 16, 22, 30, 40, 52, 66, 82], [11, 13, 17, 23, 31, 41, 53, 67, 83, 101], [12, 14, 18, 24, 32, 42, 54, 68, 84, 102, 122], [13, 15, 19, 25, 33, 43, 55, 69, 85, 103, 123, 145], [14, 16, 20, 26, 34, 44, 56, 70, 86, 104, 124, 146, 170], [15, 17, 21, 27, 35, 45, 57, 71, 87, 105, 125, 147, 171, 197], [16, 18, 22, 28, 36, 46, 58, 72, 88, 106, 126, 148, 172, 198, 226], [17, 19, 23, 29, 37, 47, 59, 73, 89, 107, 127, 149, 173, 199, 227, 257], [18, 20, 24, 30, 38, 48, 60, 74, 90, 108, 128, 150, 174, 200, 228, 258, 290], [19, 21, 25, 31, 39, 49, 61, 75, 91, 109, 129, 151, 175, 201, 229, 259, 291, 325], [20, 22, 26, 32, 40, 50, 62, 76, 92, 110, 130, 152, 176, 202, 230, 260, 292, 326, 362], [21, 23, 27, 33, 41, 51, 63, 77, 93, 111, 131, 153, 177, 203, 231, 261, 293, 327, 363, 401], [22, 24, 28, 34, 42, 52, 64, 78, 94, 112, 132, 154, 178, 204, 232, 262, 294, 32

In [13]:
reference_primes = [
    2, 3, 5, 7, 11, 13, 17, 19, 23, 29,
    31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
    73, 79, 83, 89, 97, 101, 103, 107, 109, 113,
    127, 131, 137, 139, 149, 151, 157, 163, 167, 173,
    179, 181, 191, 193, 197, 199, 211, 223, 227, 229,
    233, 239, 241, 251, 257, 263, 269, 271, 277, 281,
    283, 293, 307, 311, 313, 317, 331, 337, 347, 349,
    353, 359, 367, 373, 379, 383, 389, 397, 401, 409,
    419, 421, 431, 433, 439, 443, 449, 457, 461, 463,
    467, 479, 487, 491, 499, 503, 509, 521, 523, 541,
    547, 557, 563, 569, 571, 577, 587, 593, 599, 601,
    607, 613, 617, 619, 631, 641, 643, 647, 653, 659,
    661, 673, 677, 683, 691, 701, 709, 719, 727, 733,
    739, 743, 751, 757, 761, 769, 773, 787, 797, 809,
    811, 821, 823, 827, 829, 839, 853, 857, 859, 863,
    877, 881, 883, 887, 907, 911, 919, 929, 937, 941,
    947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013,
    1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069,
    1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151,
    1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 1223,
    1229, 1231, 1237, 1249, 1259, 1277, 1279, 1283, 1289, 1291,
    1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361, 1367, 1373,
    1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451,
    1453, 1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511,
    1523, 1531, 1543, 1549
]
filtered = [[x for x in y if x in reference_primes] for y in potentials]
print(filtered)

[[], [5, 7, 11, 17], [], [7, 13, 19, 37], [], [11, 29], [], [11, 13, 17, 23, 31, 41, 53, 67, 83, 101], [], [13, 19, 43, 103], [], [17, 71, 197], [], [17, 19, 23, 29, 37, 47, 59, 73, 89, 107, 127, 149, 173, 199, 227, 257], [], [19, 31, 61, 109, 151, 229], [], [23, 41, 131, 293, 401], [], [23, 29, 43, 53, 79, 113, 179, 233, 263, 443], [], [31, 37, 67, 97, 157, 181, 331, 367, 487, 577], [], [29, 47, 83, 137, 677], [], [29, 31, 41, 59, 71, 101, 139, 211, 239, 269, 409, 449, 491], [], [31, 37, 43, 61, 73, 103, 163, 241, 271, 337, 373, 631, 733, 787], [], [53, 89, 683], [], [37, 41, 47, 107, 167, 191, 307, 541, 587, 1091], [], [37, 43, 67, 79, 109, 127, 193, 277, 379, 457, 499, 739, 907, 967, 1093, 1297], [], [41, 59, 149, 311, 419, 1031, 1229], []]


In [14]:
more_filtered = [filtered.index(x)+4 for x in filtered if potentials[filtered.index(x)] == x and len(x) > 0] 
more_filtered # I am checking to see if the placement of the list in potentials and in filtered has not changed, and if it is not empty.

[5, 11, 17]

If you want to check your answers, here's a video on the Caboose Numbers!

https://www.youtube.com/watch?v=gM5uNcgn2NQ

### Exercise 3:
These exercises involve the use of strings. Try to predict the output in the same way as Exercise 1.

3.1: The eval() function is really good when your math expression needs to be changed rapidly. Here is the formula for projectile motion:
$$y = -\frac12gt^2+v_0t+h_0$$

As you throw an object upwards, its position (y) will depend on the time after your throw, as well as the following.
* $g$: the acceleration due to gravity (32.2 ft/s^2 or 9.8 m/s^2, depending on units)
* $v_0$: the initial upward velocity exerted by your hand
* $h_0$: the initial height (relative to some zero-level) at which the object leaves your hand

Rewrite this expression for y as a string. Remember that all instances of multiplication must be indicated with `*`. Use g, t, v, and h as the four variables. Assign this the variable `proj_motion_y`.

In [37]:
proj_motion_y = '-0.5*g*t**2+v*t+h'

Now, using only .replace(), model the equation such that the initial velocity was 5 ft/s, initial height was 9 ft, and time was 3 seconds. Assign that under any variable name different from `proj_motion_y`, and use eval() to find what the position is. 

In [41]:
e3_1a = proj_motion_y.replace('v', '5').replace('h', '9').replace('t', '3').replace('g', '32.2')
eval(e3_1a)

-120.9

Note that if you immediately (and correctly) assigned the modified `proj_motion_y` to a different variable, the original `proj_motion_y` should not have changed. Now, use .replace() and eval() to find the position when initial velocity was 13 m/s, initial height was -5 m, and time was 6 seconds.

In [44]:
e3_1b = proj_motion_y.replace('v', '13').replace('h', '-5').replace('t', '6').replace('g', '9.8')
eval(e3_1b)

-103.4

3.2: Here are a few examples of using the replace method to create new words out of 'apple'. Try making transformations of other words to make at least 10 different words. They don't all have to come from a single starting word, however.

`word_1 = 'h'+apple.replace('p', 'g')`

`word_2 = 'r'+apple.replace('p', 't')`

`word_3 = 'bed'+apple.replace('p', 'z')`

`word_4 = 's'+apple.replace('p', 'd')`

`word_5 = 'scr'+apple.replace('p', 'b')`

In [46]:
# student can do whatever words they want.
original_word = 'python'
word1 = original_word.replace('py', 'mara')
print(word1)
word2 = original_word.replace('y', 'a').replace('n', 's')
print(word2)
word3 = 'e'+original_word.replace('y', 'i').replace('on', 'et')
print(word3)
word4 = original_word.replace('py', '').replace('o', 'espia')
print(word4)
word5 = 'cr'+original_word.replace('hon', 'ic')
print(word5)
word6 = original_word.replace('t', 'rr').replace('on', 'ic')
print(word6)
word7 = original_word.replace('pyt', '')+'or'
print(word7)
word8 = original_word.replace('th', 'l')
print(word8)
word9 = original_word.replace('yt', '')+'e'
print(word9)
word10 = 'a'+original_word.replace('y','').replace('hon', '')
print(word10)

marathon
pathos
epithet
thespian
crpytic
pyrrhic
honor
pylon
phone
apt


3.3: Follow the variable names to get an output.

In [48]:
string_o_nums = '25439190243341193941203421982746422'

nine_to_four = string_o_nums.replace('9', '4')

no_four = nine_to_four.replace('4', '')

thirty_three_to_dash = no_four.replace('33', '-')

three_to_slash = thirty_three_to_dash.replace('3', '/')

no_two = three_to_slash.replace('2', '')

when = no_two
when

'5/10-11/10/1876'

3.4: Follow the variable names to get an output.

In [49]:
gibberish = 'QnR5ZmxhR3VqM2xUYWhyUG8zPQAUB3CydBml2b3IBgU3CjYB2CBzc29yARm5YdU1nbjN1Um1qbnRy'

split_A = gibberish.split('A')

middle = split_A[1]

no_B = middle.replace('B', '')

C_to_V = no_B.replace('C', 'V')

sixty_four = C_to_V
sixty_four

'U3Vydml2b3IgU3VjY2Vzc29y'

## Conclusion
Hooray! This is the end of the exercises for lab 02. I had much fun with making this one. Feel free to leave feedback below about what could be changed or improved here.

Oh right, I almost forgot: if you have found the secret code, enter it below. If you get it right, you get an additional 2 extra extra credit credit points.

<details>
    <summary>This is the solution for the secret code. Do not open if you do not want spoilers.</summary>
    There are three potential answers here, which are the (1) Philadelphia Museum of Industrial Art OR (2) Philadelphia Museum of Art OR (3) Please Touch Museum. These all come from the same two hints, those being the answers to Exercises 3.3 and 3.4. The former gives you a date range that corresponds to the 1876 Centennial Exposition in Philly. The latter is Base 64 (hence why the variable is called sixty_four) for 'Survivor Successor'. In this context, the 'Survivor' is the building that still stands from the Exposition, and the 'Successor' is that which occupied the building following the end of the Exposition. There are several such buildings, but by far the most prominent is Memorial Hall, which, after the Exposition, became the Philadelphia Museum of Industrial Art (hence (1)). That Museum would later become the Philadelphia Art Museum and move to its present location (hence (2)). After moving, the Memorial Hall became home to the Please Touch Museum in 2008 (hence (3)).
</details>